### PR TITLE
fix(geopackage overviews): create all overview levels in geopackage

### PR DESCRIPTION
**Pull request by shimoncohen**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -12,7 +12,7 @@ RUN chmod +x start.sh
 RUN python3 /app/confd/generate-config.py --environment production
 
 RUN chmod -R 777 ./confd && mkdir -p config && chmod 777 ./config && \
-    mkdir outputs && chmod -R 777 ./outputs && \
+    mkdir -p outputs && chmod -R 777 ./outputs && \
```
Comment (asafMasa) [inline Dockerfile:15]: 👍

File: src/exportImage.py [hunk 2]
```diff
@@ -132,23 +131,49 @@ def create_geopackage(self, bbox, filename, url, taskid, fullPath, resolution):
...
-            return True
-        else:
+        warp_result = gdal.Warp(fullPath, url, **kwargs)
+        self.log.info(f'Task Id "{taskid}" completed.')
+
+        if not warp_result:
             self.log.error(f'gdal return empty response for task: "{taskid}"')
             return False
 
+        Image = gdal.Open(fullPath, 1)
```
Comment (asafMasa) [inline src/exportImage.py]: refactor to method

File: src/exportImage.py [hunk 3]
```diff
@@ -132,23 +131,49 @@ def create_geopackage(self, bbox, filename, url, taskid, fullPath, resolution):
...
             'multithread': self.__config['gdal']['multithread'],
             'warpOptions': [thread_count]
         }
-        result = gdal.Warp(fullPath, url, **kwargs)
-        if result:
-            result.FlushCache()
-            result = None
-            return True
-        else:
+        warp_result = gdal.Warp(fullPath, url, **kwargs)
```
Comment (asafMasa) [inline src/exportImage.py]: refactor Warp to method
and other functionality to method

File: src/exportImage.py [hunk 4]
```diff
@@ -132,23 +131,49 @@ def create_geopackage(self, bbox, filename, url, taskid, fullPath, resolution):
...
+        self.log.info(f'Task Id "{taskid}" completed.')
+
+        if not warp_result:
             self.log.error(f'gdal return empty response for task: "{taskid}"')
             return False
 
+        Image = gdal.Open(fullPath, 1)
+        resolution_multipliers = [
+            2 ** i for i in range(1, max_zoom + 1) if self.is_valid_overview_factor(Image, 2 ** i)]
```
Comment (asafMasa) [inline src/exportImage.py]: add unit test
explain the method

File: src/exportImage.py [hunk 5]
```diff
@@ -167,3 +192,22 @@ def runPreperation(self, taskId, filename):
...
+        top_right_lat = bbox[3]
+        top_right_lon = bbox[2]
+        bottom_left_lat = bbox[1]
+        bottom_left_lon = bbox[0]
+
+        # Check if bbox width and height are at least at the resolution of a pixle at the wanted zoom level
+        return (top_right_lon - bottom_left_lon) > float(resolution) and (top_right_lat - bottom_left_lat) > float(resolution)
+
+    def is_valid_overview_factor(self, dataset, ov_factor):
+        return dataset.RasterXSize / ov_factor >= 8 and dataset.RasterYSize / ov_factor >= 8
```
Comment (asafMasa) [inline src/exportImage.py]: why to use get_zoom_resolution if resolution data already exists in the image data

**Review by asafMasa — CHANGES_REQUESTED** [review]
👍 Good Job

File: src/exportImage.py [hunk 6]
```diff
@@ -10,28 +11,46 @@
...
+    return dataset.RasterXSize / ov_factor >= 8 and dataset.RasterYSize / ov_factor >= 8
+
+
+def calculate_overviews(dataframe, max_zoom):
+    """
+    Calculate overview factors for the given dataframe.
+
+    return an array of valid overview factors (2 ^ i) for the dataset
+    """
+    return [2 ** i for i in range(1, max_zoom + 1) if is_valid_overview_factor(dataframe, 2 ** i)]
```
Comment (asafMasa) [inline src/exportImage.py]: separate the return from the result - first calculate then return the result

File: src/exportImage.py [hunk 7]
```diff
@@ -121,49 +147,92 @@ def delete_local_directory(self, directoryName):
...
+            bbox, filename, url, task_id, full_path, resolution)
+
+        if not warp_result:
+            self.log.error(f'gdal return empty response for task: "{task_id}"')
+            return False
+
+        overviews_result = self.create_geopackage_overview(
+            filename, task_id, full_path, max_zoom)
+
+        # Clear cache (for closing process)
```
Comment (asafMasa) [inline src/exportImage.py:185]: should'nt it be before the "create_geopackage_overview" method?

File: src/exportImage.py [hunk 8]
```diff
@@ -121,49 +147,92 @@ def delete_local_directory(self, directoryName):
...
+        warp_result.FlushCache()
+        warp_result = None
+
+        # Check for overview build errors
+        if overviews_result == GDALBuildOverviewsResponse.CE_Failure or overviews_result == GDALBuildOverviewsResponse.CE_Fatal:
+            self.log.error(
+                f'gdal return empty response for task: "{task_id}" (overviews)')
```
Comment (asafMasa) [inline src/exportImage.py]: its not empty, why not log the "overviews_result "?

File: src/exportImage.py [hunk 9]
```diff
@@ -121,49 +147,92 @@ def delete_local_directory(self, directoryName):
...
-        result = gdal.Warp(fullPath, url, **kwargs)
-        if result:
-            result.FlushCache()
-            result = None
-            return True
-        else:
-            self.log.error(f'gdal return empty response for task: "{taskid}"')
-            return False
+        result = gdal.Warp(full_path, url, **kwargs)
+        self.log.info(f'Task Id "{task_id}" completed.')
```
Comment (asafMasa) [inline src/exportImage.py]: it's not completed as the log said - only first stage is completed (geopackage creation)

File: tests/exportImage.py [hunk 10]
```diff
@@ -44,6 +46,14 @@ def test_invalid_export(self):
         result = is_valid_zoom_for_area(resolution, bbox)
         self.assertEqual(result, False)
 
+    # Check that correct overviews are calculated
+    def test_calculate_overviews_correctly(self):
+        fullPath = f'{path.dirname(__file__)}/test.GPKG'
+        Image = gdal.Open(fullPath, 1)
+        overviews = calculate_overviews(Image, 18)
+        # Should have overviews 17, ..., 9
+        self.assertEqual(overviews, [2, 4, 8, 16, 32, 64, 128, 256, 512])
```
Comment (asafMasa) [inline tests/exportImage.py:55]: shouldn't be more?

File: src/exportImage.py [hunk 11]
```diff
@@ -10,28 +11,46 @@
...
+    CE_Warning = 2
+    CE_Failure = 3
+    CE_Fatal = 4
 
 
 class ExportImage:
     def __init__(self):
         self.log = Logger.get_logger_instance()
         self.__helper = Helper()
+        self.warp_percent = 70
```
Comment (asafMasa) [inline src/exportImage.py:58]: do you think to leave 1% for the index creation and actual s3 upload?

File: src/exportImage.py [hunk 12]
```diff
@@ -121,49 +147,92 @@ def delete_local_directory(self, directoryName):
...
+
+        # Build overviews
+        self.log.info(f'Creating overviews for task {task_id}')
+        es_obj = {"taskId": task_id, "filename": filename}
+        kwargs = {
+            'callback': self.overviews_progress_callback,
+            'callback_data': es_obj
+        }
+        overviews_result = Image.BuildOverviews(
+            "BILINEAR", resolution_multipliers, **kwargs)
```
Comment (asafMasa) [inline src/exportImage.py:249]: check with @shaziri that "BILINEAR" is the desired method

Comment (asafMasa) [reply src/exportImage.py:249]: check if "Multithreading" option is supported

**Review by asafMasa — CHANGES_REQUESTED** [review]
thanks

File: tests/exportImage.py (hunk 10 above)
Comment (shimoncohen) [reply tests/exportImage.py:55]: No, not for the size of the original raster

File: src/exportImage.py (hunk 8 above)
Comment (shimoncohen) [reply src/exportImage.py]: Added

File: src/exportImage.py (hunk 7 above)
Comment (shimoncohen) [reply src/exportImage.py:185]: You are correct

File: src/exportImage.py (hunk 9 above)
Comment (shimoncohen) [reply src/exportImage.py]: Moved and corrected log

File: src/exportImage.py (hunk 6 above)
Comment (shimoncohen) [reply src/exportImage.py]: Changed

File: src/exportImage.py (hunk 11 above)
Comment (shimoncohen) [reply src/exportImage.py:58]: Added

**Review by asafMasa — CHANGES_REQUESTED** [review]
Great Job - just remove the "tests/test.GPKG " file

**Review by asafMasa — APPROVED** [review]

**Pull request merged**